From cec43272d9debeaf6294606001cbf540cc3c34a5 Mon Sep 17 00:00:00 2001
From: Patrick Combes <patrick.combes@intel.com>
Date: Tue, 3 Sep 2013 17:21:49 +0200
Subject: [PORT FROM MAIN][BT] Improve robustness of GKI_add_to_timer_list

BZ: 135318

It has been seen that GKI_add_to_timer_list may be called with
p_timer_listq->p_first == NULL although p_timer_listq->last_ticks > 0

This should not happen however in case of p_timer_listq->p_first == NULL, we
should consider that the list is empty whatever p_timer_listq->last_ticks
value.

Change-Id: If74227ff3db07c4f7d28ab6af3798981056f2cb4
Orig-Change-Id: I382bef0decda4f7ccf3431cd1bf5ff5b2248bf94
Category: device enablement
Domain: CWS.BT-Common
Origin: internal
Upstream-Candidate: yes
Signed-off-by: Patrick Combes <patrick.combes@intel.com>
Reviewed-by: Gros, Julien <julien.gros@intel.com>
Tested-by: Lestienne, ChristopheX <christophex.lestienne@intel.com>
Reviewed-on: http://android.intel.com:8080/131384
Reviewed-by: Lestienne, ChristopheX <christophex.lestienne@intel.com>
---
 gki/common/gki_time.c |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/gki/common/gki_time.c b/gki/common/gki_time.c
index a9af8fa..a48d496 100644
--- a/gki/common/gki_time.c
+++ b/gki/common/gki_time.c
@@ -817,8 +817,8 @@ void GKI_add_to_timer_list (TIMER_LIST_Q *p_timer_listq, TIMER_LIST_ENT  *p_tle)
     /* Only process valid tick values */
     if (p_tle->ticks >= 0)
     {
-        /* If this entry is the last in the list */
-        if (p_tle->ticks >= p_timer_listq->last_ticks)
+        /* If this entry is the only entry or the last in the list */
+        if (p_timer_listq->p_first == NULL || p_tle->ticks >= p_timer_listq->last_ticks)
         {
             /* If this entry is the only entry in the list */
             if (p_timer_listq->p_first == NULL)
-- 
1.7.4.1

